### PR TITLE
raft: don't panic when looking for term conflicts

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This commit fixes a hypothetical panic that may occur when a stale `MsgApp`
message arrives to a follower. The conflict searching algorithm in
`findConflictByTerm` may return a log index which is not present in the log, and
thus the `raftLog.term()` method may return an error. It is safe to ignore this
error and send `MsgAppResp` with the found index and a zero `LogTerm`.

This commit also restores the behaviour that existed before commit d0fb0cd6.
Back then, the `term()` function would silently return 0 instead of an error, and
a zero `LogTerm` would be sent with the rejecting `MsgAppResp`. After that commit,
there is a new possible panic. We remove the possibility of this panic here.

### CODE REVIEW

File: log.go [hunk 1]
```diff
@@ -182,12 +189,13 @@ func (l *raftLog) findConflictByTerm(index uint64, term uint64) uint64 {
 			index, li)
 		return index
 	}
-	for {
+	for ; ; index-- {
 		logTerm, err := l.term(index)
 		if logTerm <= term || err != nil {
+			// NB: an error can only be ErrCompacted here, but it's always safe to
+			// return the current index, so we don't assert on other errors.
```
Comment (tbg) [inline log.go:196]: Yes and no because the method comment claims that this returns the "best possible" index but in the case of a transient storage error this isn't true. Could we return an error here if we see one that is unexpected (i.e. not ErrCompacted).

File: raft.go [hunk 2]
```diff
@@ -1667,9 +1667,13 @@ func (r *raft) handleAppendEntries(m pb.Message) {
 		hintIndex := min(m.Index, r.raftLog.lastIndex())
 		hintIndex = r.raftLog.findConflictByTerm(hintIndex, m.LogTerm)
 		hintTerm, err := r.raftLog.term(hintIndex)
-		if err != nil {
-			panic(fmt.Sprintf("term(%d) must be valid, but got %v", hintIndex, err))
+		if err != nil { // can be ErrCompacted or ErrUnavailable
```
Comment (tbg) [inline raft.go:1670]: something like `panicUnlessErrCompactedOrErrUnavailable(err)`?

File: raft.go [hunk 3]
```diff
@@ -1667,9 +1667,13 @@ func (r *raft) handleAppendEntries(m pb.Message) {
 		hintIndex := min(m.Index, r.raftLog.lastIndex())
 		hintIndex = r.raftLog.findConflictByTerm(hintIndex, m.LogTerm)
```
Comment (tbg) [inline raft.go:1668]: Wouldn't it be better for this to return the term, since we're just about to recompute it?

File: raft.go [hunk 4]
```diff
@@ -1667,9 +1667,13 @@ func (r *raft) handleAppendEntries(m pb.Message) {
...
 		hintTerm, err := r.raftLog.term(hintIndex)
-		if err != nil {
-			panic(fmt.Sprintf("term(%d) must be valid, but got %v", hintIndex, err))
+		if err != nil { // can be ErrCompacted or ErrUnavailable
+			hintTerm = 0
 		}
+		// NB: findConflictByTerm always returns an index <= its argument, so
+		// hintIndex <= min(m.Index, lastIndex()). This guarantees that the next
+		// append will be at a lower index. This is also the reason why it is safe
+		// to ignore a possible error above which, though, should never happen.
```
Comment (tbg) [inline raft.go:1676]: I'm still uneasy about not populating the term hint. The reason the leader can handle this is because originally it was never populated, but we should consider this legacy behavior we'd want to migrate out of if we had a way to do so. If `findConflictByTerm` returned `(index, term, error)` and on success both `(index,term)` are populated, there would be fewer questions. And on error, we could log a warning and simply not respond, as is admissible in raft, and in particular for a rejection, or we could respond with the unadjusted `(index,term)`, so something like this:

```go
hintIndex := ...
hintTerm, err := ....
if err != nil {
  return // don't respond, log warning
}
adjHintIndex, adjHintTerm, err := r.raftLog.findConflictByTerm(...)
if err != nil { log warning but continue } else {
hintIndex, hintTerm = adjHintIndex, adjHintTerm
}
// respond
```

**Review by tbg — APPROVED** [review]
Looks good to me, thanks! I left some suggestions to a have fewer "unexpected" return values and no case in which the term is not populated. The goal would be to remove all edge cases and make it more transparent what happens in the truncation scenarios, etc.

It might also be good to add some targeted unit testing for `findConflictByTerm`. I saw that you added a test case in `TestFastLogRejection`, but a few more wouldn't hurt. I should have added that right when we introduced `findConflictByTerm`, sorry about that.

Approving preemptively because the current code is "good enough" and I'm fairly certain it's correct (mod comments). The improvements would make it easier to be convinced that it's correct in a month or two but I'll rely on your judgment of which further improvements are worth it.

File: log.go [hunk 5]
```diff
@@ -164,15 +164,22 @@ func (l *raftLog) findConflict(ents []pb.Entry) uint64 {
...
+// log l with index <= `index` and term <= `term`.
 //
 // The index provided MUST be equal to or less than l.lastIndex(). Invalid
 // inputs log a warning and the input index is returned.
+//
+// If index < firstIndex()-1 then the information about the term at this index
+// has been compacted away. In this case the input index is returned. Similarly,
+// if all terms for entries in [firstIndex()-1, index] are greater than the
+// given term, i.e. the possible suitable entry has been compacted, then
+// firstIndex()-2 is returned. Note that these two cases probably indicate a
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply log.go]: We would return 0 still, because `raftLog.term(0) == 0`. But I refactored this func to guard against `index` underflow, and also made the method comment better.

File: log.go [hunk 6]
```diff
@@ -182,12 +189,13 @@ func (l *raftLog) findConflictByTerm(index uint64, term uint64) uint64 {
 			index, li)
 		return index
 	}
-	for {
+	for ; ; index-- {
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply log.go]: With the next iteration of this PR, this loop is now `for ; index > 0; index-- {` which is slightly more justified. Leaving as is.

File: log.go (hunk 1 above)
Comment (pavelkalinnikov) [reply log.go]: This comment is softer now. Decided not to return an err / overengineer here, because `term()` method currently panics on any error other than `ErrCompacted/ErrUnavailable`. This all feels clunky and not ideal, maybe error handling in this library should be improved.

But also: returning both a higher and a lower hint here is fine - if it doesn't work, there will be another MsgApp iteration, or there will be more entries resent than necessary. The `MsgApp` sender (leader) works in such a way that it can swallow any hint including zero.

File: raft.go (hunk 3 above)
Comment (pavelkalinnikov) [reply raft.go]: Done.

File: raft.go (hunk 2 above)
Comment (pavelkalinnikov) [reply raft.go]: Moved away from this error check.

File: raft.go (hunk 4 above)
Comment (pavelkalinnikov) [reply raft.go]: As discussed in chat, it's more important to send a rejection `MsgAppResp` than get the hint exactly right.

There is an implicit expectation that a flow of `MsgApp` messages will eventually lead to a `MsgResp` (potentially with a reject). This is sort of a liveness bump for appends. If we introduce a condition under which a rejection `MsgResp` is not sent, we add some risk of ending up in a busy loop on the leader.

File: raft.go [hunk 7]
```diff
@@ -1909,6 +1909,8 @@ func (r *raft) abortLeaderTransfer() {
 
 // committedEntryInCurrentTerm return true if the peer has committed an entry in its term.
 func (r *raft) committedEntryInCurrentTerm() bool {
+	// NB: r.Term is never 0 on a leader, so if zeroTermOnOutOfBounds returns 0,
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply raft.go:1914]: Yes, I manually checked that this func is only called on a leader. But it's also somewhat intuitive: the leader checks that it committed entries during its current term.

File: raft_test.go [hunk 8]
```diff
@@ -4703,72 +4705,52 @@ func TestFastLogRejection(t *testing.T) {
...
-				t.Errorf("peer 1 step append fail")
-			}
+			require.NoError(t, n1.Step(msgs[0]), "peer 1 step append fail")
 			msgs = n1.readMessages()
-			if msgs[0].LogTerm != test.nextAppendTerm {
-				t.Fatalf("#%d expected log term = %d, but got %d", i, test.nextAppendTerm, msgs[0].LogTerm)
-			}
-			if msgs[0].Index != test.nextAppendIndex {
-				t.Fatalf("#%d expected index = %d, but got %d", i, test.nextAppendIndex, msgs[0].Index)
-			}
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply raft_test.go:4771]: Removed.

**Comment by pavelkalinnikov** [comment]
@tbg Addressed comments, and added unit tests for `findConflictByTerm`.

**Comment by pavelkalinnikov** [comment]
CC @ahrtr: This removes a panic condition. Please merge since this is now approved, or feel free to review too.

@tbg I modified this PR after approval, but it's equivalent. Feel free to take a final look though, and please merge if you see this earlier.

Thank you.
